server/finder/handler/finder_handler.go

```diff
@@ -56,40 +57,60 @@ func (h *FinderHandler) Find(mhashes []multihash.Multihash) (*model.FindResponse
 
 		provResults := make([]model.ProviderResult, 0, len(values))
 		for j := range values {
-			provID := values[j].ProviderID
-			// Lookup provider info for each unique provider, look in local map
-			// before going to registry.
-			addrs, ok := provAddrs[provID]
+			iVal := values[j]
+			provID := iVal.ProviderID
+			addrs := h.fetchProviderAddrs(provID, iVal.ContextID, provAddrs, true)
+
+			if addrs == nil {
+				continue
+			}
+
+			// Check if there is an extended provider record for that provID
+			epRecord, ok := extendedProviders[provID]
 			if !ok {
-				pinfo, allowed := h.registry.ProviderInfo(provID)
-				if pinfo == nil {
-					// If provider not in registry, then provider was deleted.
-					// Tell the indexed core to delete the contextID for the
-					// deleted provider. Delete the contextID from the core,
-					// because there is no way to delete all records for the
-					// provider without a scan of the entire core valuestore.
-					go func(value indexer.Value) {
-						err := h.indexer.RemoveProviderContext(value.ProviderID, value.ContextID)
-						if err != nil {
-							log.Errorw("Error removing provider context", "err", err)
-						}
-					}(values[j])
-					// If provider not in registry, do not return in result.
-					continue
-				}
-				// Omit provider info if not allowed or marked as inactive.
-				if !allowed || pinfo.Inactive() {
-					continue
-				}
-				addrs = pinfo.AddrInfo.Addrs
-				provAddrs[provID] = addrs
+				epRecord = h.registry.ExtendedProviders(provID)
```

Comment:
If would be more efficient to get the provider addrs and the extended provider addrs from the registry at the same time, otherwise you have to enter the critical goroutine twice for each response.  Same would be true of a critical mutex. This could cause contention which is already a concern with a high read load.

Reply:
Done as a result of moving `ExtendedProviders` into `ProviderInfo`